kubernetes-maven-plugin/plugin/src/main/java/org/eclipse/jkube/maven/plugin/mojo/build/ApplyMojo.java

```diff
@@ -186,55 +178,18 @@ public void executeInternal() throws MojoExecutionException {
             KubernetesResourceUtil.validateKubernetesMasterUrl(masterUrl);
             log.info("Using %s at %s in namespace %s with manifest %s ", clusterKind, masterUrl, clusterAccess.getNamespace(), manifest);
 
-            applyService.setAllowCreate(createNewResources);
-            applyService.setServicesOnlyMode(servicesOnly);
-            applyService.setIgnoreServiceMode(ignoreServices);
-            applyService.setLogJsonDir(jsonLogDir);
-            applyService.setBasedir(getRootProjectFolder());
-            applyService.setIgnoreRunningOAuthClients(ignoreRunningOAuthClients);
-            applyService.setProcessTemplatesLocally(processTemplatesLocally);
-            applyService.setDeletePodsOnReplicationControllerUpdate(deletePodsOnReplicationControllerUpdate);
-            applyService.setRollingUpgrade(rollingUpgrades);
-            applyService.setRollingUpgradePreserveScale(isRollingUpgradePreserveScale());
-            applyService.setRecreateMode(recreate);
-
-            boolean openShift = OpenshiftHelper.isOpenShift(kubernetes);
-            if (openShift) {
-                log.info("[[B]]OpenShift[[B]] platform detected");
-            } else {
-                disableOpenShiftFeatures(applyService);
-            }
+            configureApplyService(kubernetes);
 
             Set<HasMetadata> entities = KubernetesResourceUtil.loadResources(manifest);
-
+            // Configure Namespace: Check if new namespace needs to be created and create if
+            // it's provided in manifests first. Then use that namespace for applying rest of
+            // the manifests
             String namespace = clusterAccess.getNamespace();
-            boolean namespaceEntityExist = false;
-
-            for (HasMetadata entity: entities) {
-                if (entity instanceof Namespace) {
-                    Namespace ns = (Namespace)entity;
-                    namespace = ns.getMetadata().getName();
-                    applyService.applyNamespace((ns));
-                    namespaceEntityExist = true;
-                    entities.remove(entity);
-                    break;
-                }
-                if (entity instanceof Project) {
-                    Project project = (Project)entity;
-                    namespace = project.getMetadata().getName();
-                    applyService.applyProject(project);
-                    namespaceEntityExist = true;
-                    entities.remove(entity);
-                    break;
-                }
-            }
-
-            if (!namespaceEntityExist) {
-                applyService.applyNamespace(namespace);
-            }
-
+            applyService.applyNamespaceOrProjectIfPresent(entities);
```

Comment:
This is weird and unclear.
Could you explain what's going on?
We retrieve the namespace, but the namespace is only set/configured later on in the method.
In addition, the invocation in line 188 (`applyService.applyNamespaceOrProjectIfPresent`) suggests we are already doing something here, but we only configure the namespace later on.
There's also a call `configureApplyService` in line 181. It feels that the namespace part should be really configured there.

My understanding is that if the namespace is configured later, is because it has something to do with the processing and it affects the end result. 
If this is the case `setNamespace` should be completely removed and shifted to an argument of any method that performs processing.

Reply:
I think this code handles the case when there is a `Namespace` present in `entities`. I think It handles the case when you want your resources to be created in a new namespace which would be created by jkube during apply phase. Namespace gets created first, then resources are applied in that particular namespace.

okay, let me shift `setNamespace` to `configureApplyService`